clang/tools/driver/cc1depscand_main.cpp

```diff
@@ -703,12 +748,13 @@ static void updateCompilerInvocation(CompilerInvocation &Invocation,
                                      llvm::TreePathPrefixMapper &Mapper) {
   // Fix the CAS options.
   auto &CASOpts = Invocation.getCASOpts();
-  CASOpts.Kind = CASOptions::BuiltinCAS; // FIXME: Don't override.
+  if (CASOpts.Kind == CASOptions::NoCAS) {
+    CASOpts.Kind = CASOptions::BuiltinCAS;
+    CASOpts.BuiltinPath = llvm::cas::getDefaultOnDiskCASStableID();
+  }
```

Comment:
Something this is potentially making harder to see is that the path to the CAS is going to show up in the `-cc1`, which then becomes part of the cache key.

`getDefaultOnDiskCASStableID()` is a string that is always the same. It's okay for it to be part of the CAS key IMO. That's exactly why the function exists.

But, I believe, the actual path on disk to the CAS should somehow be excluded from the cache key. 

I had been planning to "solve" that before allowing a CAS path to be passed in. Any ideas for how to solve it? If you think this important to land, ideas for where to put a FIXME describing the problem?

Reply:
> the path to the CAS is going to show up in the -cc1, which then becomes part of the cache key.

Ah good point, I'll take another look with that in mind.

Reply:
One idea:
- Restrict CASOptions to things that should NOT be part of the CAS key (maybe `-fcas-token-cache` doesn't belong here then?)
- Temporarily set CASOptions to default-constructed when generating the `-cc1` for the CAS key (or add a `CompilerInvocation` function that skips them)

I worry this could potentially hide problems with secondary outputs, like diagnostics. And how do we audit to make sure CASOptions doesn't accidentally pick up something that does affect the output?

Another idea:
- Remove CASOptions from CompilerInvocation. Instead require it to be parsed separately (could still use marshalling for parsing / etc... could make it easier to share with the driver too?)
- Maybe leave behind a hollowed out `CASDB` somewhere that's owned by CompilerInvocation that isn't tied to a command-line option

The nice thing about this is that CompilerInvocation stays clean (just serialize the whole thing!). But I think this creates other problems...

Reply:
> Restrict CASOptions to things that should NOT be part of the CAS key

I like this, we could have a separate `CASDBOptions` so that configuration for DB creation a separate thing than how it gets used.

Reply:
Probably these should move over to FileSystemOptions anyway:
```
  /// If set, uses this root ID with \a CASFileSystem.
  std::string CASFileSystemRootID;

  /// If set, used as the working directory for -fcas-fs.
  std::string CASFileSystemWorkingDirectory;
```
they're probably(?) only in `CASOptions` because I was trying to isolate changes I made to other parts of the code.